### PR TITLE
Implementation of HTTP/2 fuzzer

### PR DESCRIPTION
**Pull request by alegz-filomaphitskay**
Contributes to #1297

What was done here:
- added implementation of HTTP/2 fuzzer;
- added more explicit definition of buffer overflow that is storing generated message.

### CODE REVIEW

File: fw/t/fuzzer.c [hunk 1]
```diff
@@ -50,8 +50,13 @@ typedef struct {
 static FuzzMsg spaces[] = {
```
Comment (krizhanovsky) [inline fw/t/fuzzer.c:50]: Please update the copyright year to 2015-2022 and the module version to 0.2.0

File: fw/t/fuzzer.c [hunk 2]
```diff
@@ -353,36 +373,87 @@ gen_vector_move(TfwFuzzContext *ctx, int i)
...
+{
+	BUG_ON(!ch);
+
+	if(*p == end) {
+		ctx->is_buf_not_enough = true;
+	}
+	else
+	{
+		*(*p)++ = ch;
+	}
```
Comment (krizhanovsky) [inline fw/t/fuzzer.c:412]:
```suggestion
	if(*p == end)
		ctx->is_buf_not_enough = true;
	else
		*(*p)++ = ch;
```

File: fw/t/fuzzer.c [hunk 3]
```diff
@@ -353,36 +373,87 @@ gen_vector_move(TfwFuzzContext *ctx, int i)
...
+add_string_n(TfwFuzzContext *ctx, char **p, char *end, const char *str, int n)
+{
+	while (*p != end && n) {
+		*(*p)++ = *str++;
+		--n;
+	}
+
+	if (n) {
+		ctx->is_buf_not_enough = true;
+	}
```
Comment (krizhanovsky) [inline fw/t/fuzzer.c:425]:
```suggestion
	if (n)
		ctx->is_buf_not_enough = true;
```

File: fw/t/fuzzer.c [hunk 4]
```diff
@@ -353,36 +373,87 @@ gen_vector_move(TfwFuzzContext *ctx, int i)
...
 {
-	for (; *str != '\0'; str++)
-		addch(p, end, *str);
+	while (*p != end && *str != '\0') {
+		*(*p)++ = *str++;
+	}
+
+	if (*str != '\0') {
+		ctx->is_buf_not_enough = true;
+	}
```
Comment (krizhanovsky) [inline fw/t/fuzzer.c:437]: Here and in other places: we don't use braces for single line statements. BTW we use the coding style, inherited from the kernel, https://github.com/tempesta-tech/tempesta/blob/master/CodingStyle

File: fw/t/fuzzer.c [hunk 5]
```diff
@@ -453,24 +524,55 @@ add_field(TfwFuzzContext *ctx, int type, char **p, char *end, int t)
...
+	if (type == FUZZ_REQ_H2) {
+		saved_p = *p;
+		saved_end = end;
+		remain_p = saved_end - saved_p;
+		BUG_ON(remain_p > sizeof(h2_buf));
+		*p = h2_buf;
+		end = h2_buf + remain_p;
+	}
+	else
+	{
```
Comment (krizhanovsky) [inline fw/t/fuzzer.c:546]:
```suggestion
	} else {
```
also see the coding style

File: fw/t/fuzzer.c [hunk 6]
```diff
@@ -353,36 +373,87 @@ gen_vector_move(TfwFuzzContext *ctx, int i)
...
+	}
+}
+
+static void
+add_string_n(TfwFuzzContext *ctx, char **p, char *end, const char *str, int n)
+{
+	while (*p != end && n) {
+		*(*p)++ = *str++;
+		--n;
+	}
```
Comment (krizhanovsky) [inline fw/t/fuzzer.c:421]: why not memcpy()?

File: fw/t/fuzzer.c [hunk 7]
```diff
@@ -353,36 +373,87 @@ gen_vector_move(TfwFuzzContext *ctx, int i)
...
 
 static void
-add_string(char **p, char *end, const char *str)
+add_string(TfwFuzzContext *ctx, char **p, char *end, const char *str)
 {
-	for (; *str != '\0'; str++)
-		addch(p, end, *str);
+	while (*p != end && *str != '\0') {
+		*(*p)++ = *str++;
+	}
```
Comment (krizhanovsky) [inline fw/t/fuzzer.c:433]: strncpy()?

**Review by krizhanovsky — APPROVED** [review]
LGTM after minor cleanups

File: fw/t/fuzzer.c (hunk 4 above)
Comment (alegz-filomaphitskay) [reply fw/t/fuzzer.c:437]: Done

File: fw/t/fuzzer.c (hunk 5 above)
Comment (alegz-filomaphitskay) [reply fw/t/fuzzer.c:546]: Done

File: fw/t/fuzzer.c (hunk 6 above)
Comment (alegz-filomaphitskay) [reply fw/t/fuzzer.c:417]: Seemed so much elegant). I changed the code to use memcpy.

File: fw/t/fuzzer.c [hunk 8]
```diff
@@ -3,11 +3,11 @@
  *
  * Tempesta HTTP fuzzer.
  *
- * Copyright (C) 2015-2017 Tempesta Technologies, Inc.
+ * Copyright (C) 2015-2022 Tempesta Technologies, Inc.
  *
  * This program is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License as published by
- * the Free Software Foundation; either version 2 of the License,
+ * the Free Software Foundation; either version 0.2.0 of the License,
```
Comment (krizhanovsky) [inline fw/t/fuzzer.c:10]: No, the version of the module in line 32, not the GPL version :)

**Review by krizhanovsky — APPROVED** [review]
LGTM: just do the cleanup and merge it after #1597. There is no need to next review

File: fw/t/fuzzer.c (hunk 8 above)
Comment (alegz-filomaphitskay) [reply fw/t/fuzzer.c:10]: Shame on me :(

**Pull request merged**